lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,236 @@
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+require 'set'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # Files present in :files are not processed further. The others
+      #   are added to the Hash. This behavior is the same implemented in openstack-infra/shade
+      #   see https://github.com/openstack-infra/shade/blob/1d16f64fbf376a956cafed1b3edd8e51ccc16f2c/shade/openstackcloud.py#L1200
+
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :template_ori
+        attr_reader :visited
+        attr_reader :max_files_size
+
+        def initialize(template, files = nil)
+          @template = template
+          @template_ori = template
+          @template_base_url = nil
+          @files = files || {}
+          @max_files_size = (128 * 1 << 10)
+          @visited = {}
+        end
+
+        def yaml_load(content)
+          return YAML.load(content) if RUBY_VERSION < "2.1"
+
+          YAML.safe_load(content, [Date])
+        end
+
+        def get_files
+          Fog::Logger.warning("Processing template #{@template}")
+          _, @template = get_template_contents(@template)
```

Comment:
if @template is a String, `get_template_contents` can't modify it. We need to always replace it there.

Removing the assignment, the proper unittest fails.

```
Fog::Orchestration[:openstack] | stack requests::success#test_0005_#dont_modify_passed_template [/code/test/requests/orchestration/stack_files_util_tests.rb:89]:

      # The template argument should be modified.
      assert(template['resources']['a_file']['type'].start_with?('file:///'), file_resolver.template)
      ^^^ this will fail.

      # No side effect on the original template.
      refute(@local_yaml['resources']['a_file']['type'].start_with?('file:///'))

```